adapters/amx/amx_test.go

```diff
@@ -0,0 +1,354 @@
+package amx
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/openrtb_ext"
+
+	"github.com/prebid/prebid-server/adapters/adapterstest"
+)
+
+var (
+	bidRequest string
+)
+
+const (
+	amxTestEndpoint = "http://pbs-dev.amxrtb.com/auction/openrtb"
+	defaultImpExt   = "{\"bidder\":{\"tagId\":\"publisher_id_example\"}}"
+)
+
+func TestJsonSamples(t *testing.T) {
+	adapterstest.RunJSONBidderTest(t, "amxtest", new(AMXAdapter))
+}
+
+func TestMakeRequestsPublisherIdOverride(t *testing.T) {
+	var w, h int = 300, 250
+
+	var width, height uint64 = uint64(w), uint64(h)
+
+	adapter := NewAMXBidder(amxTestEndpoint)
+	imp1 := openrtb.Imp{
+		ID:  "sample_imp_1",
+		Ext: json.RawMessage(defaultImpExt),
+		Banner: &openrtb.Banner{
+			W: &width,
+			H: &height,
+			Format: []openrtb.Format{
+				{W: 300, H: 250},
+			},
+		}}
+
+	inputRequest := openrtb.BidRequest{
+		User: &openrtb.User{ID: "example_user_id"},
+		Imp:  []openrtb.Imp{imp1},
+		Site: &openrtb.Site{
+			Publisher: &openrtb.Publisher{
+				ID: "1234567",
+			},
+		},
+		ID: "1234",
+	}
+
+	actualAdapterRequests, err := adapter.MakeRequests(&inputRequest, &adapters.ExtraRequestInfo{})
+
+	if actualAdapterRequests == nil {
+		t.Errorf("request should be nil")
+	}
+
+	if len(err) != 0 {
+		t.Errorf("We should have no error")
+	}
+
+	// check that the publisher ID overrides the site.publisher.id if provided
+	var body openrtb.BidRequest
+	if err := json.Unmarshal(actualAdapterRequests[0].Body, &body); err != nil {
+		t.Errorf("failed to read bid request")
+	}
+	if body.Site.Publisher.ID != "publisher_id_example" {
+		t.Errorf("invalid publisher id: %s", body.Site.Publisher.ID)
+	}
+}
+
+func TestMakeRequestsApp(t *testing.T) {
+	var w, h int = 300, 250
+
+	var width, height uint64 = uint64(w), uint64(h)
+
+	adapter := NewAMXBidder(amxTestEndpoint)
+	imp1 := openrtb.Imp{
+		ID:  "sample_imp_1",
+		Ext: json.RawMessage("{\"bidder\":{\"tagId\":\"site_publisher_id\"}}"),
+		Banner: &openrtb.Banner{
+			W: &width,
+			H: &height,
+			Format: []openrtb.Format{
+				{W: 300, H: 250},
+			},
+		}}
+
+	inputRequest := openrtb.BidRequest{
+		Imp: []openrtb.Imp{imp1},
+		Site: &openrtb.Site{
+			Publisher: &openrtb.Publisher{
+				ID: "1234567",
+			},
+		},
+		App: &openrtb.App{ID: "cansanuabnua", Publisher: &openrtb.Publisher{ID: "whatever"}},
+		ID:  "1234",
+	}
```

Comment:
It's not valid for a BidRequest to contain both a `Site` and `App` section. This will be caught earlier on by the framework. For this test, please only use the `App`.